src/crypto/CHIPCryptoPALOpenSSL.cpp

```diff
@@ -725,53 +598,83 @@ CHIP_ERROR P256PublicKey::ECDSA_validate_hash_signature(const uint8_t * hash, co
 
     VerifyOrExit(hash != nullptr, error = CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrExit(hash_length == kSHA256_Hash_Length, error = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(signature.Length() == kP256_ECDSA_Signature_Length_Raw, error = CHIP_ERROR_INVALID_ARGUMENT);
+
     nid = _nidForCurve(MapECName(Type()));
     VerifyOrExit(nid != NID_undef, error = CHIP_ERROR_INVALID_ARGUMENT);
 
     ec_group = EC_GROUP_new_by_curve_name(nid);
-    VerifyOrExit(ec_group != nullptr, error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(ec_group != nullptr, error = CHIP_ERROR_NO_MEMORY);
 
     key_point = EC_POINT_new(ec_group);
-    VerifyOrExit(key_point != nullptr, error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(key_point != nullptr, error = CHIP_ERROR_NO_MEMORY);
 
     result = EC_POINT_oct2point(ec_group, key_point, Uint8::to_const_uchar(*this), Length(), nullptr);
     VerifyOrExit(result == 1, error = CHIP_ERROR_INTERNAL);
 
     ec_key = EC_KEY_new_by_curve_name(nid);
-    VerifyOrExit(ec_key != nullptr, error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(ec_key != nullptr, error = CHIP_ERROR_NO_MEMORY);
 
     result = EC_KEY_set_public_key(ec_key, key_point);
     VerifyOrExit(result == 1, error = CHIP_ERROR_INTERNAL);
 
     result = EC_KEY_check_key(ec_key);
     VerifyOrExit(result == 1, error = CHIP_ERROR_INTERNAL);
 
-    // The cast for length arguments is safe because values are small enough to fit.
-    result = ECDSA_verify(0, hash, static_cast<int>(hash_length), Uint8::to_const_uchar(signature),
-                          static_cast<int>(signature.Length()), ec_key);
+    // Build-up the signature object from raw <r,s> tuple
+    r = BN_bin2bn(Uint8::to_const_uchar(signature.ConstBytes()) + 0u, kP256_FE_Length, nullptr);
+    VerifyOrExit(r != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+    s = BN_bin2bn(Uint8::to_const_uchar(signature.ConstBytes()) + kP256_FE_Length, kP256_FE_Length, nullptr);
+    VerifyOrExit(s != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+    ec_sig = ECDSA_SIG_new();
+    VerifyOrExit(ec_sig != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+    result = ECDSA_SIG_set0(ec_sig, r, s);
+    VerifyOrExit(result == 1, error = CHIP_ERROR_INTERNAL);
+
+    result = ECDSA_do_verify(Uint8::to_const_uchar(hash), static_cast<int>(hash_length), ec_sig, ec_key);
     VerifyOrExit(result == 1, error = CHIP_ERROR_INVALID_SIGNATURE);
     error = CHIP_NO_ERROR;
 
 exit:
     _logSSLError();
-    if (ec_group != nullptr)
+    if (ec_sig != nullptr)
     {
-        EC_GROUP_free(ec_group);
-        ec_group = nullptr;
+        ECDSA_SIG_free(ec_sig);
+
+        // After ECDSA_SIG_set0 succeeds, r and s memory is managed by ECDSA_SIG object.
+        // We set to nullptr so that we don't try to double-free
+        r = nullptr;
+        s = nullptr;
     }
-    if (key_point != nullptr)
+    if (s != nullptr)
     {
-        EC_POINT_clear_free(key_point);
-        key_point = nullptr;
+        BN_clear_free(s);
```

Comment:
ScopedBigNum and ScopedPoint would be so nice. I think some OpenSSL C++ headers have unique_ptr implementations for this.